inline/src/parser/mod.rs

```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
-pub struct Parser {
+pub struct Parser<'tokens> {
     /// Iterator over [`Token`]s found in Unimarkup input.
     ///
     /// [`Token`]: crate::Token
-    iter: Tokens,
+    iter: Tokens<'tokens>,
 
     /// Storage of [`Token`] already yielded from [`TokenIterator`] but not consumed in current
     /// iteration of parsing.
     ///
     /// [`Token`]: crate::Token
     /// [`TokenIterator`]: crate::TokenIterator
     /// [`Inline`]: crate::Inline
-    token_cache: Option<Token>,
+    token_cache: Option<Token<'tokens>>,
 
     /// Storage of parsed [`Inline`]s that should be returned before parsing next [`Inline`].
     inline_cache: VecDeque<Inline>,
 }
 
-impl Parser {
+impl<'tokens> Parser<'tokens> {
     /// Returns the next [`Token`] either from [`Lexer`] directly or from internal token cache.
     ///
     /// [`Token`]: crate::Token
     /// [`Lexer`]: crate::Lexer
-    fn next_token(&mut self) -> Option<Token> {
+    fn next_token(&mut self) -> Option<Token<'tokens>> {
         if self.token_cache.is_some() {
             self.token_cache.take()
         } else {
             self.iter.next()
         }
     }
 
-    fn parse_inline(&mut self, token: Token) -> Inline {
-        // opening/closing of tokens is resolved at lexing stage
-        // at this point we can simply parse
-        let kind = token.kind;
-        let start = token.span.start;
-        let mut end = token.span.end;
+    fn parse_plain(&mut self, start_token: Token, cxt: PlainContext) -> Inline {
+        let kind = if start_token.is_plain_whitespace() {
+            TokenKind::Plain
+        } else {
+            start_token.kind
+        };
 
-        if kind != TokenKind::Plain && !token.opens() {
-            let (content, span) = token.into_inner();
-            let content = InlineContent::Plain(PlainContent::new(content, span));
-            return Inline::as_plain_or_eol(content, kind);
-        }
+        let (tkn_str, mut span) = start_token.parts();
 
-        let mut content: InlineContent<_, _> = NestedContent::default().into();
+        let mut content = if cxt.enclosed {
+            // skip first (the enclosing) token
+            String::new()
+        } else {
+            String::from(tkn_str)
+        };
 
-        if kind == TokenKind::Plain {
-            content.append(InlineContent::from(token));
+        while let Some(next_token) = self.next_token() {
+            let enclosed_and_closes = cxt.enclosed && next_token.closes(Some(&start_token));
+            let not_enclosed_and_interrupted = !cxt.enclosed && next_token.kind != kind;
+
+            if enclosed_and_closes {
+                span.end = next_token.span.end;
+                break;
+            } else if not_enclosed_and_interrupted {
+                if next_token.is_plain_whitespace() {
+                    // consume whitespace
+                    let (next_content, next_span) = next_token.parts();
+                    content.push_str(next_content);
```

Comment:
We do check already with `is_whitespace` in scanner (see `symbol` impl in `unimarkup_commons` crate). I agree about pushing plain space 👍🏻

Reply:
This is also a potential one where edge cases can be solved with a function on Token. Something like: `fn content_for(&self, inline: &Inline) -> &str`. The changes like `consumable_by_plain` function etc. make this a little tedious to handle right now, but since it's the single edge case for now I would keep a little tedious but easy to understand for now.